Bug fix for forward refs in generics

mark-todd opened the pull request:
## Change Summary


Converts string to Forward Refs to allow for generic case
Also to fix discriminated unions allows Literals to pass through replace types unchanged.

## Related issue number

fix #6130


Selected Reviewer: @dmontagu

dmontagu commented:
I see this is breaking one of the discriminated union tests, do you have an idea of how to resolve that? If we can get that fixed I'd be happy to merge this.

mark-todd commented:
Hi David - so I've been trying to figure this out, this much I know:

The test used to fail for me in python3.10, then I added the check for is Literal (it was failing because the string in the literal was being converted to a ForwardRef)

I fixed it, but then when I uploaded I found it was producing the same bug in python3.9 and below, but not 3.10 or 3.11. I don't understand this because it should use the same function. It must be something to do with the conditional imports based on version that are changing the behaviour. I'll look into it some more and see what I can find but any thoughts on it would be appreciated!

mark-todd commented:
I've had a follow up thought on this - 3.9 or less does not have the list dict features, so it's possible there is a different part of the code that executes instead of replace args but that has a similar role.

mark-todd commented:
@dmontagu All tests fixed - should be good to merge
